### PR TITLE
 [TF] optimization cont

### PR DESCRIPTION
**Pull request by abhinavsp0730**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

Notebooks covered.
1. momentum. 
2.adagrad.

### CODE REVIEW

File: chapter_optimization/adagrad.md [hunk 1]
```diff
@@ -139,13 +147,28 @@ def adagrad(params, states, hyperparams):
...
+```{.python .input}
+#@tab tensorflow
+def init_adagrad_states(feature_dim):
+    s_w = tf.Variable(d2l.zeros((feature_dim, 1)))
+    s_b = tf.Variable(d2l.zeros(1))
+    return (s_w, s_b)
+
+def adagrad(params, grads, states, hyperparams):
+    eps = 1e-6
+    for p, s, g in zip(params, states, grads):
```
Comment (abhinavsp0730) [inline chapter_optimization/adagrad.md:159]: There is a bug over here, please help if you guys can figure it. @terrytangyuan @astonzhang

**Comment by mli** [comment]
Job d2l-en/PR-1282/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1282/

File: chapter_optimization/adagrad.md (hunk 1 above)
Comment (abhinavsp0730) [reply chapter_optimization/adagrad.md:160]: Fixed.

**Comment by mli** [comment]
Job d2l-en/PR-1282/2 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1282/

**Comment by mli** [comment]
Job d2l-en/PR-1282/3 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1282/

File: chapter_optimization/lr-scheduler.md [hunk 2]
```diff
@@ -92,13 +155,29 @@ class SquareRootScheduler:
...
 
 ```{.python .input}
 scheduler = SquareRootScheduler(lr=1.0)
 d2l.plot(np.arange(num_epochs), [scheduler(t) for t in range(num_epochs)])
 ```
 
+```{.python .input}
+#@tab tensorflow
+schedular = SquareRootScheduler(1.0)
+d2l.plot(d2l.arange(num_epochs), [schedular(t) for t in range(num_epochs)])
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:178]:
```suggestion
scheduler = SquareRootScheduler(1.0)
d2l.plot(d2l.arange(num_epochs), [scheduler(t) for t in range(num_epochs)])
```

File: chapter_optimization/lr-scheduler.md [hunk 3]
```diff
@@ -107,6 +186,12 @@ trainer = gluon.Trainer(net.collect_params(), 'sgd',
 train(net, train_iter, test_iter, num_epochs, loss, trainer, device)
 ```
 
+```{.python .input}
+#@tab tensorflow
+train(net, train_iter, test_iter, num_epochs, lr,
+      custom_callback=LearningRateScheduler(schedular))
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:192]:
```suggestion
      custom_callback=LearningRateScheduler(scheduler))
```

File: chapter_optimization/lr-scheduler.md [hunk 4]
```diff
@@ -152,6 +272,12 @@ trainer = gluon.Trainer(net.collect_params(), 'sgd',
...
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:278]:
```suggestion
      custom_callback=LearningRateScheduler(scheduler))
```

File: chapter_optimization/lr-scheduler.md [hunk 5]
```diff
@@ -166,6 +292,49 @@ scheduler = lr_scheduler.CosineScheduler(max_update=20, base_lr=0.5,
...
+      self.base_lr = self.final_lr + (self.base_lr_orig - self.final_lr) * \
+                (1 + math.cos(math.pi * (epoch - self.warmup_steps) /
+                              self.max_steps)) / 2
+    return self.base_lr
+
+
+scheduler = CosineScheduler(max_update=20, base_lr=0.5,
+                                         final_lr=0.01)
+
+d2l.plot(d2l.arange(num_epochs), [schedular(t) for t in range(num_epochs)])
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:335]:
```suggestion
d2l.plot(d2l.arange(num_epochs), [scheduler(t) for t in range(num_epochs)])
```

File: chapter_optimization/lr-scheduler.md [hunk 6]
```diff
@@ -174,6 +343,12 @@ trainer = gluon.Trainer(net.collect_params(), 'sgd',
 train(net, train_iter, test_iter, num_epochs, loss, trainer, device)
 ```
 
+```{.python .input}
+#@tab tensorflow
+train(net, train_iter, test_iter, num_epochs, lr,
+      custom_callback=LearningRateScheduler(schedular))
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:349]:
```suggestion
      custom_callback=LearningRateScheduler(scheduler))
```

File: chapter_optimization/lr-scheduler.md [hunk 7]
```diff
@@ -194,6 +376,12 @@ trainer = gluon.Trainer(net.collect_params(), 'sgd',
...
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:382]:
```suggestion
      custom_callback=LearningRateScheduler(scheduler))
```

File: chapter_optimization/lr-scheduler.md [hunk 8]
```diff
@@ -63,6 +63,54 @@ def train(net, train_iter, test_iter, num_epochs, loss, trainer, device):
...
+    return tf.keras.models.Sequential([
+        tf.keras.layers.Conv2D(filters=6, kernel_size=5, activation='sigmoid',
+                               padding='same'),
+        tf.keras.layers.AvgPool2D(pool_size=2, strides=2),
+        tf.keras.layers.Conv2D(filters=16, kernel_size=5,
+                               activation='sigmoid'),
+        tf.keras.layers.AvgPool2D(pool_size=2, strides=2),
+        tf.keras.layers.Flatten(),
+        tf.keras.layers.Dense(120, activation='sigmoid'),
+        tf.keras.layers.Dense(84, activation='sigmoid'),
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:84]:
```suggestion
        tf.keras.layers.Dense(120, activation='relu'),
        tf.keras.layers.Dense(84, activation='relu'),
```

File: chapter_optimization/lr-scheduler.md [hunk 9]
```diff
@@ -63,6 +63,54 @@ def train(net, train_iter, test_iter, num_epochs, loss, trainer, device):
...
+#@tab tensorflow
+%matplotlib inline
+from d2l import tensorflow as d2l
+import tensorflow as tf
+import math
+from tensorflow.keras.callbacks import LearningRateScheduler
+
+def net():
+    return tf.keras.models.Sequential([
+        tf.keras.layers.Conv2D(filters=6, kernel_size=5, activation='sigmoid',
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:76]:
```suggestion
        tf.keras.layers.Conv2D(filters=6, kernel_size=5, activation='relu,
```

File: chapter_optimization/lr-scheduler.md [hunk 10]
```diff
@@ -63,6 +63,54 @@ def train(net, train_iter, test_iter, num_epochs, loss, trainer, device):
...
+                               padding='same'),
+        tf.keras.layers.AvgPool2D(pool_size=2, strides=2),
+        tf.keras.layers.Conv2D(filters=16, kernel_size=5,
+                               activation='sigmoid'),
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:80]:
```suggestion
                               activation='relu'),
```

File: chapter_optimization/lr-scheduler.md [hunk 11]
```diff
@@ -63,6 +63,54 @@ def train(net, train_iter, test_iter, num_epochs, loss, trainer, device):
...
+        net.compile(optimizer=optimizer, loss=loss, metrics=['accuracy'])
+    callback = d2l.TrainCallback(net, train_iter, test_iter, num_epochs,
+                             device_name)
+    if custom_callback is False:
+        net.fit(train_iter, epochs=num_epochs, verbose=0, 
+                callbacks=[callback])
+    else:
+         net.fit(train_iter, epochs=num_epochs, verbose=0,
+                 callbacks=[callback, custom_callback])
+    return net
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:111]: why is the trainer/optimizer and loss defined within the function?
If possible please follow MXNet style so that there is some sort of consistency.

File: chapter_optimization/lr-scheduler.md [hunk 12]
```diff
@@ -144,6 +245,25 @@ scheduler = lr_scheduler.MultiFactorScheduler(step=[15, 30], factor=0.5,
...
+    self.factor = factor
+    self.base_lr = base_lr
+  
+  def __call__(self, epoch):
+    if epoch in range(self.step[0], (self.step[1] + 1)):
+      return self.base_lr * self.factor
+    else:
+      return self.base_lr
+
+scheduler = MultiFactorScheduler(step=[15, 30], factor=0.5,base_lr=0.5)
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:262]:
```suggestion
scheduler = MultiFactorScheduler(step=[15, 30], factor=0.5, base_lr=0.5)
```

File: chapter_optimization/adagrad.md [hunk 13]
```diff
@@ -139,11 +147,26 @@ def adagrad(params, states, hyperparams):
...
+        s[:].assign(s + tf.math.square(g))
+        p[:].assign(p - hyperparams['lr'] * g / tf.math.sqrt(s + eps))
+```
+
 Compared to the experiment in :numref:`sec_minibatch_sgd` we use a
 larger learning rate to train the model.
 
 ```{.python .input}
 #@tab all
+# todo fix adagrad
```
Comment (AnirudhDagar) [inline chapter_optimization/adagrad.md:169]: Is this fixed?

File: chapter_optimization/lr-scheduler.md (hunk 11 above)
Comment (abhinavsp0730) [reply chapter_optimization/lr-scheduler.md:111]: It's required, So that the model will be train in the GPU. We've to compile the model inside tf.distribute.Strategy scope.

File: chapter_optimization/adagrad.md (hunk 13 above)
Comment (abhinavsp0730) [reply chapter_optimization/adagrad.md:169]: Yes, it's fixed. Sorry, actually I forget to remove it.

File: chapter_optimization/lr-scheduler.md (hunk 8 above)
Comment (abhinavsp0730) [reply chapter_optimization/lr-scheduler.md]: Let the second last layer's activation be sigmoid. If not then some LR scheduler will fail to converge

Comment (AnirudhDagar) [reply chapter_optimization/lr-scheduler.md]: @astonzhang

**Comment by mli** [comment]
Job d2l-en/PR-1282/9 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1282/

**Comment by abhinavsp0730** [comment]
@terrytangyuan please review.

File: chapter_optimization/adadelta.md [hunk 14]
```diff
@@ -83,6 +83,28 @@ def adadelta(params, states, hyperparams):
...
+%matplotlib inline
+from d2l import tensorflow as d2l
+import tensorflow as tf
+
+def init_adadelta_states(feature_dim):
+    s_w = tf.Variable(d2l.zeros((feature_dim, 1)))
+    s_b = tf.Variable(d2l.zeros(1))
+    delta_w = tf.Variable(d2l.zeros((feature_dim, 1)))
+    delta_b = tf.Variable(d2l.zeros(1))
+    return ((s_w, delta_w), (s_b, delta_b))
```
Comment (terrytangyuan) [inline chapter_optimization/adadelta.md:97]: Remove redundant parenthesis

File: chapter_optimization/adagrad.md [hunk 15]
```diff
@@ -139,6 +147,20 @@ def adagrad(params, states, hyperparams):
         p.grad.data.zero_()
 ```
 
+```{.python .input}
+#@tab tensorflow
+def init_adagrad_states(feature_dim):
+    s_w = tf.Variable(d2l.zeros((feature_dim, 1)))
+    s_b = tf.Variable(d2l.zeros(1))
+    return (s_w, s_b)
```
Comment (terrytangyuan) [inline chapter_optimization/adagrad.md:155]: The parenthesis is redundant

File: chapter_optimization/lr-scheduler.md [hunk 16]
```diff
@@ -81,6 +135,15 @@ trainer.set_learning_rate(0.1)
 print(f'learning rate is now {trainer.learning_rate:.2f}')
 ```
 
+```{.python .input}
+#@tab tensorflow
+lr = 0.1
+dummy_model = tf.keras.models.Sequential([tf.keras.layers.Dense(10)])
+dummy_model.compile(tf.keras.optimizers.SGD(learning_rate=lr), loss='mse')
+
+print(f'learning rate is now ,', dummy_model.optimizer.lr.numpy())
```
Comment (AnirudhDagar) [inline chapter_optimization/lr-scheduler.md:144]: Please use f strings the right way.
```suggestion
print(f'learning rate is now {dummy_model.optimizer.lr.numpy():.2f}')
```

File: chapter_optimization/adadelta.md (hunk 14 above)
Comment (abhinavsp0730) [reply chapter_optimization/adadelta.md:97]: @terrytangyuan Are you talking about this ```((s_w, delta_w), (s_b, delta_b))``` parenthesis. If so then this parenthesis is used in both mxnet and pytorch.

File: chapter_optimization/adagrad.md (hunk 15 above)
Comment (abhinavsp0730) [reply chapter_optimization/adagrad.md:155]: @terrytangyuan Are you talking about this ```(s_w, s_b)``` parenthesis. If so then this parenthesis is used in both mxnet and pytorch.

Comment (terrytangyuan) [reply chapter_optimization/adagrad.md:155]: Oh I see then let's keep it consistent for now and polish later.

File: chapter_optimization/lr-scheduler.md (hunk 16 above)
Comment (abhinavsp0730) [reply chapter_optimization/lr-scheduler.md:144]: Fixed

**Comment by abhinavsp0730** [comment]
@terrytangyuan please review now.

File: chapter_optimization/momentum.md [hunk 17]
```diff
@@ -139,13 +154,28 @@ Let us see how momentum works in practice, i.e., when used within the context of
...
     v_b = d2l.zeros(1)
     return (v_w, v_b)
 ```
 
+```{.python .input}
+#@tab pytorch
+def init_momentum_states(feature_dim):
+    v_w = d2l.zeros((feature_dim, 1))
+    v_b = d2l.zeros(1)
+    return (v_w, v_b)
```
Comment (astonzhang) [inline chapter_optimization/momentum.md:168]: use @tab mxnet,pytorch

File: chapter_optimization/rmsprop.md [hunk 18]
```diff
@@ -77,13 +84,28 @@ d2l.show_trace_2d(f_2d, d2l.train_2d(rmsprop_2d))
...
     s_b = d2l.zeros(1)
     return (s_w, s_b)
 ```
 
+```{.python .input}
+#@tab pytorch
+def init_rmsprop_states(feature_dim):
+    s_w = d2l.zeros((feature_dim, 1))
+    s_b = d2l.zeros(1)
+    return (s_w, s_b)
```
Comment (astonzhang) [inline chapter_optimization/rmsprop.md:98]: use `#@tab mxnet,pytorch`

Comment (abhinavsp0730) [reply chapter_optimization/rmsprop.md:98]: @astonzhang do I need to commit changes in this PR or should I open a new one?

Comment (AnirudhDagar) [reply chapter_optimization/rmsprop.md:98]: They are now fixed in master.
